includes/class-plugin-manager.php

```diff
@@ -0,0 +1,270 @@
+<?php
+/**
+ * Newspack setup
+ *
+ * @package Newspack
+ */
+
+namespace Newspack;
+
+defined( 'ABSPATH' ) || exit;
+
+/**
+ * General purpose class for managing installation/activation of plugins.
+ */
+class Plugin_Manager {
+
+	/**
+	 * Determine whether plugin installation is allowed in the current environment.
+	 *
+	 * @return bool
+	 */
+	public static function can_install_plugins() {
+		if ( ( defined( 'DISALLOW_FILE_EDIT' ) && DISALLOW_FILE_EDIT ) || ( defined( 'DISALLOW_FILE_MODS' ) && DISALLOW_FILE_MODS ) ) {
+			return false;
+		}
+
+		return true;
+	}
+
+	/**
+	 * Activate a plugin, installing it first if necessary.
+	 *
+	 * @param string $plugin The plugin slug or URL to the plugin.
+	 * @return bool True on success. False on failure or if plugin was already activated.
+	 */
+	public static function activate( $plugin ) {
+		if ( ! function_exists( 'get_plugins' ) ) {
+			require_once ABSPATH . 'wp-admin/includes/plugin.php';
+		}
+
+		$plugin_slug = $plugin;
+
+		// if it's a url, parse the slug from the URL.
+		if ( wp_http_validate_url( $plugin ) ) {
+			$plugin_slug = self::get_plugin_slug_from_url( $plugin );
+			if ( ! $plugin_slug ) {
+				return false;
+			}
+		}
+
+		$installed_plugins = self::get_installed_plugins();
+
+		// Install the plugin if it's not installed already.
+		$plugin_installed = isset( $installed_plugins[ $plugin_slug ] );
+		if ( ! $plugin_installed ) {
+			$plugin_installed = self::install( $plugin );
+		}
+		if ( ! $plugin_installed ) {
+			return false;
+		}
+
+		// Refresh the installed plugin list if the plugin isn't present because we just installed it.
+		if ( ! isset( $installed_plugins[ $plugin_slug ] ) ) {
+			$installed_plugins = self::get_installed_plugins();
+		}
+
+		if ( is_plugin_active( $installed_plugins[ $plugin_slug ] ) ) {
+			return false;
+		}
+
+		$activated = activate_plugin( $installed_plugins[ $plugin_slug ] );
+		if ( is_wp_error( $activated ) ) {
+			return false;
+		}
+
+		return true;
+	}
+
+	/**
+	 * Deactivate a plugin.
+	 *
+	 * @param string $plugin_file The path to the plugin file. e.g. 'newspack/newspack.php'.
+	 * @return bool True on success. False on failure.
+	 */
+	public static function deactivate( $plugin_file ) {
+		$installed_plugins = self::get_installed_plugins();
+		if ( ! in_array( $plugin_file, $installed_plugins ) ) {
+			return false;
+		}
+
+		deactivate_plugins( $plugin_file );
+		return true;
+	}
+
+	/**
+	 * Get a simple list of all installed plugins.
+	 *
+	 * @return array of 'plugin_slug => plugin_file_path' entries for all installed plugins.
+	 */
+	public static function get_installed_plugins() {
+		$installed_plugins = array_reduce( array_keys( get_plugins() ), array( __CLASS__, 'reduce_plugin_info' ) );
+		if ( empty( $installed_plugins ) ) {
+			$installed_plugins = [];
+		}
+		return $installed_plugins;
+	}
+
+	/**
+	 * Parse a plugin slug from the URL to download a plugin.
+	 *
+	 * @param string $url The URL to a plugin zip file.
```

Comment:
Could we be clever about this and allow anything that we can determine is a WP.org plugin URL and work out the actual ZIP URL from that?

Reply:
We could; it's just going to be e.g. `woo-gutenberg-products-block` for `https://wordpress.org/plugins/woo-gutenberg-products-block/`. I don't know if that's going to be a feature we need, since we already know the plugin slug for all .org plugins (it's `woo-gutenberg-products-block` for the plugin I just mentioned).

I'll add it if needed in a future PR, though. Good thinking and very feasible. :)